op-program/client/interop/consolidate.go

```diff
@@ -0,0 +1,227 @@
+
+	for i, chain := range superRoot.Chains {
+		progress := transitionState.PendingProgress[i]
+
+		// TODO(13776): hint block data execution in case the pending progress is not canonical so we can fetch the correct receipts
+		block, receipts := oracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
+		execMsgs, _, err := ReceiptsToExecutingMessages(deps.DependencySet(), receipts)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+
+		candidate := supervisortypes.BlockSeal{
+			Hash:      progress.BlockHash,
+			Number:    block.NumberU64(),
+			Timestamp: block.Time(),
+		}
+		hazards, err := getHazards(progress.BlockHash, execMsgs)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+		if err := checkHazards(deps, &candidate, hazards); err != nil {
+			// TODO(13776): replace with deposit-only block if ErrConflict, ErrCycle, or ErrFuture
+			return eth.Bytes32{}, err
+		}
+		consolidatedChains = append(consolidatedChains, eth.ChainIDAndOutput{
+			ChainID: chain.ChainID,
+			// TODO(13776): when applicable, use the deposit-only block output root
+			Output: progress.OutputRoot,
+		})
+	}
+	consolidatedSuper := &eth.SuperV1{
+		Timestamp: superRoot.Timestamp + 1,
+		Chains:    consolidatedChains,
+	}
+	return eth.SuperRoot(consolidatedSuper), nil
+}
+
+type ConsolidateCheckDeps interface {
+	cross.UnsafeFrontierCheckDeps
+	cross.CycleCheckDeps
+}
+
+func checkHazards(
+	deps ConsolidateCheckDeps,
+	candidate *supervisortypes.BlockSeal,
+	hazards map[supervisortypes.ChainIndex]supervisortypes.BlockSeal,
+) error {
+	// TODO: for each of these hazard checks, ensure that they're scoped to the super root timestamp of the chain
+	if err := cross.HazardUnsafeFrontierChecks(deps, hazards); err != nil {
+		return err
+	}
+	if err := cross.HazardCycleChecks(deps.DependencySet(), deps, candidate.Timestamp, hazards); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getHazards treats all executing messages as hazards
+func getHazards(
+	blockHash common.Hash,
+	execMsgs []*supervisortypes.ExecutingMessage,
+) (map[supervisortypes.ChainIndex]supervisortypes.BlockSeal, error) {
+	hazards := make(map[supervisortypes.ChainIndex]supervisortypes.BlockSeal)
+	for _, execMsg := range execMsgs {
+		hazards[execMsg.Chain] = supervisortypes.BlockSeal{
+			Hash:      blockHash,
+			Number:    execMsg.BlockNum,
+			Timestamp: execMsg.Timestamp,
+		}
+	}
+	return hazards, nil
+}
+
+type consolidateCheckDeps struct {
+	oracle      l2.Oracle
+	depset      depset.DependencySet
+	canonBlocks map[uint64]*l2.CanonicalBlockHeaderOracle
+}
+
+func newConsolidateCheckDeps(chains []eth.ChainIDAndOutput, oracle l2.Oracle) (*consolidateCheckDeps, error) {
+	// TODO: handle case where dep set changes in a given timestamp
+	// TODO: Also replace dep set stubs with the actual dependency set in the RollupConfig.
+	deps := make(map[eth.ChainID]*depset.StaticConfigDependency)
+	canonBlocks := make(map[uint64]*l2.CanonicalBlockHeaderOracle)
```

Comment:
These two things seem unrelated. They're both iterating over the chains, but two distinct collections are created. Can you split it into two separate helper functions?

Reply:
Split into two separate for loops so it's easier to read. Can use helper functions if you prefer.